app/src/main/java/com/example/weather/ui/MainActivity.kt

```diff
@@ -1,20 +1,157 @@
+
+        mFusedLocationClient.lastLocation
+            .addOnSuccessListener { location: Location? ->
+                mLastLocation = location
+                initWeatherView(location)
+            }
+
+        PermissionUtils.getLastLocation(
+            this,
+            this,
+            PermissionUtils.isLocationEnabled(this)
+        )
+    }
+
+    override fun onPause() {
+        unregisterNetworkBroadcastReceiver(this)
+        super.onPause()
+    }
+
+    override fun onRestart() {
+        super.onRestart()
+        initWeatherView(mCurrentLocation)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        PermissionUtils.onRequestPermissionResult(
+            requestCode,
+            grantResults,
+            this
+        )
+    }
+
+    override fun onDataLocation(location: Location?) {
+        this.mCurrentLocation = location
+        val distance = mCurrentLocation?.let { currentLocation ->
+            mLastLocation?.let { lastLocation ->
+                Utils.distanceBetweenPoints(
+                    currentLocation.latitude,
+                    currentLocation.longitude,
+                    lastLocation.latitude,
+                    lastLocation.longitude
+                )
+            }
+        }
+        if (distance != null) {
+            if (distance > Constant.MIN_DISTANCE_FIRST_TRIGGER) {
+                initWeatherView(location)
+            }
+        } else {
+            initWeatherView(location)
+        }
+    }
+
+    override fun networkAvailable() {
+        sharedViewModel.checkNetwork(true)
+    }
+
+    override fun networkUnavailable() {
+        Toast.makeText(
+            this,
+            getString(R.string.message_network_not_responding),
+            Toast.LENGTH_SHORT
+        ).show()
+        sharedViewModel.checkNetwork(false)
+    }
+
+    private fun onLocationRequest() {
+        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
+    private fun requestPermissions() {
+        if (!PermissionUtils.checkPermissions(this)) {
+            PermissionUtils.requestPermissions(this)
+        }
+    }
+
+    private fun initWeatherView(location: Location?) {
+        location?.let { location ->
+            addFragmentToActivity(
+                supportFragmentManager,
+                WeatherFragment.newInstance(location.latitude, location.longitude),
+                R.id.container
+            )
+        }
+    }
+
+    private fun startNetworkBroadcastReceiver(currentContext: Context) {
+        networkStateReceiver = NetworkStateReceiver()
+        networkStateReceiver!!.addListener((currentContext as NetworkStateReceiverListener))
```

Comment:
```suggestion
if(currentContext is NetworkStateReceiverListener) {
        networkStateReceiver.addListener((currentContext as NetworkStateReceiverListener))
  }
```
Không force cast như này sẽ crash app nếu currentContex không implement NetworkStateReceiverListener